### PR TITLE
Prevent arbitrary column renaming in the schema update

### LINKED ISSUE #646: Fehlerhafte Vorschläge zur Datenbankaktualisierung im Installtool
Nach der Installation von Contao 4.4.42 mit Newsletter-Bundle und Kopieren der Datenbank aus Contao 3.5.40 werden im Installtool zur Aktualisierung der Datenbank in der standardmäßig vorausgewählten Rubrik "Spalten ändern" u. a. die folgenden beiden Änderungen vorgeschlagen:
````
ALTER TABLE tl_newsletter_channel CHANGE smtpuser sender VARCHAR(128) DEFAULT '' NOT NULL
ALTER TABLE tl_newsletter_channel CHANGE smtppass template VARCHAR(32) DEFAULT '' NOT NULL
````
Statt die Spalten ```sender``` und  ```template``` neu anzulegen, werden die Spalten ```smtpuser``` und ```smtppass``` umbebannt.

### LINKED ISSUE #1716: Database update leads to fields being renamed which can't be renamed
**Affected version(s)**

4.4

**Description**

In Contao 4 (I guess it's not been this way in v3) the database install tool added the feature that if a field is scheduled to be deleted and there's one to be added, the tool suggests to `CHANGE` the field which is definitely a problem in some cases because the field content is copied also. This can lead to errors when some value is in there which is interpreted by the logic for the new field.

Is this intended behavior?

Bye Defcon0

### PR DESCRIPTION
**Pull request by leofeyer**
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #646 #1716
| Docs PR or issue | -

Solution provided by @dmolineus. Thank you.

### CODE REVIEW

**Comment by bytehead** [comment]
can this behaviour be configured somewhere?

**Comment by leofeyer** [comment]
No, it was never configurable and it is not now. Do you need it to be configurable? Because as the issue title says, the renaming is arbitrary.

**Comment by ausi** [comment]
So if I understand that correctly, we never want rely on the renaming detection of doctrine. If the core or an extension renames a column it has to provide a proper migration for it?

**Comment by bytehead** [comment]
It is not configurable now, yes. But it renames columns now (in 80% of my cases is renaming more convenient because I'll not lose any data e.g. during development).

**Comment by bytehead** [comment]
Or is it possible to activate this only for Contao prefixed tables? I'm not a fan of altering this process for all my applications just because Contao don't want to rename anymore.

**Comment by leofeyer** [comment]
> So if I understand that correctly, we never want rely on the renaming detection of doctrine. If the core or an extension renames a column it has to provide a proper migration for it?

Yes.

> It is not configurable now, yes. But it renames columns now (in 80% of my cases is renaming more convenient because I'll not lose any data e.g. during development).

It tried to rename `tl_member.trustedtokenversion` to `tl_member.createdOn` on my system. This would lead to corrupt data and members who have signed up on January 1st, 1970! Is this really what you want?

**Comment by bytehead** [comment]
This is not a real world use case. But imagine there's just a typo somewhere. Totally safe to rename.

**Comment by dmolineus** [comment]
As I stated in the doctrine ticket, it's impossible for the machine to know if a rename fits. It would be great if a user could decide for each rename statement if he want's to. If it's worth to to develop such solution for the install tool and the migrate command, that's the question.

**Comment by ausi** [comment]
> Or is it possible to activate this only for Contao prefixed tables?

As we want to remove the table prefix, I don’t think this would be a proper solution.

**Comment by ausi** [comment]
> can this behaviour be configured somewhere?

Would removing the `contao.listener.doctrine_listener` in your projects _services.yml_ work?

**Comment by bytehead** [comment]
> As we want to remove the table prefix, I don’t think this would be a proper solution.

Then I don't think this is a solution for outside the Contao universum. The installation bundle will change doctrine's default behaviour for all the application. This has to be configurable somewhere.

**Comment by bytehead** [comment]
> Would removing the contao.listener.doctrine_listener in your projects services.yml work?

Funny, I tried that yesterday with the `contao.listener.doctrine_schema`, but it didn't work and the event listener was still called on `postGenerateSchema`. So I guess not?

**Review by aschempp — CHANGES_REQUESTED** [review]
I don't think manually writing the SQL is correct. We should use the `Platform::getAlterTableSQL()` method instead.

**Comment by leofeyer** [comment]
> This is not a real world use case.

This is a 100% real world use case, because it occurred in one of my installations. 🤷‍♂️

**Comment by leofeyer** [comment]
As discussed in Mumble on June 4th, we want to move the listener to the manager bundle.

**Review by aschempp — CHANGES_REQUESTED** [review]
> I don't think manually writing the SQL is correct. We should use the Platform::getAlterTableSQL() method instead.

That hasn't changed?

**Comment by leofeyer** [comment]
I don't think you are right. 😄 The `getAlterTableSQL()` method returns one giant statement to execute all ALTER TABLE changes at once. Not what we need.

**Comment by aschempp** [comment]
why not? You're currently executing two `ALTER TABLE` statements, the platform method would combine these into just one statement?

File: manager-bundle/src/EventListener/DoctrineListener.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ */
+
+namespace Contao\ManagerBundle\EventListener;
+
+use Doctrine\DBAL\Event\SchemaAlterTableRenameColumnEventArgs;
+
+/**
+ * @internal
+ */
+class DoctrineListener
```
Comment (aschempp) [inline manager-bundle/src/EventListener/DoctrineListener.php]: I also don't really like this name, don't we already have several doctrine listeners that do different things? Should this be a `DoctrineAlterTableListener` ?

**Comment by leofeyer** [comment]
I really do not see how this causes anything but extra work, but anyway: 081009406e44e317da3652ecd970d9d6cb3de554

File: manager-bundle/src/EventListener/DoctrineListener.php (hunk 1 above)
Comment (leofeyer) [reply manager-bundle/src/EventListener/DoctrineListener.php]: Renamed in eb4acea2b951aa7545a38a72aa4120e1f880dec7

**Review by Toflar — APPROVED** [review]
LGTM. Nice work 👍

File: manager-bundle/src/EventListener/DoctrineAlterTableListener.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        $args->preventDefault();
+
+        $platform = $args->getPlatform();
+        $table = $args->getTableDiff()->getName($platform);
+        $oldColumn = $args->getTableDiff()->fromTable->getColumn($args->getOldColumnName());
+        $column = $args->getColumn();
+
+        $tableDiff = new TableDiff($table->getName());
+        $tableDiff->removedColumns[$args->getOldColumnName()] = $oldColumn;
```
Comment (aschempp) [inline manager-bundle/src/EventListener/DoctrineAlterTableListener.php:35]: You could probably use the new `$column` here as well as Doctrine only uses the key, but not sure. Not important either :D

**Review by aschempp — APPROVED** [review]

**Review by ausi — APPROVED** [review]

**Pull request merged**